ecl/hqlcpp/hqlhtcpp.cpp

```diff
@@ -10993,35 +11054,55 @@ ABoundActivity * HqlCppTranslator::doBuildActivityOutput(BuildCtx & ctx, IHqlExp
         return doBuildActivityOutputWorkunit(ctx, expr, isRoot);
 
     OwnedHqlExpr filename = foldHqlExpression(rawFilename);
-    IHqlExpression * program  = queryRealChild(expr, 2);
+
+    IHqlExpression * fileTypeOptionsExpr  = queryRealChild(expr, 2); // PIPE() or TYPE() on output
+    StringBuffer genericFileTypeFormat;
+    IHqlExpression * pipe = NULL;
+    if (fileTypeOptionsExpr)
+    {
+        if (fileTypeOptionsExpr->getOperator() == no_filetype)
+        {
+            // Force use of generic I/O if we're using file type plugins;
+            // assign the format based on the file type.
+            useGenericReadWrites = true;
+            IHqlExpression * fileType = queryAttributeChild(fileTypeOptionsExpr, fileTypeAtom, 0);
+            getStringValue(genericFileTypeFormat, fileType);
+            genericFileTypeFormat.toLowerCase();
+        }
+        else if (fileTypeOptionsExpr->getOperator() == no_pipe)
+        {
+            pipe = fileTypeOptionsExpr->queryChild(0);
+            useGenericReadWrites = false;
+        }
+    }
+    else if (filename->getOperator() == no_pipe)
+    {
+        pipe = filename->queryChild(0);
+        useGenericReadWrites = false;
+    }
+
+    LinkedHqlExpr expireAttr = expr->queryAttribute(expireAtom);
+    IHqlExpression * seq = querySequence(expr);
+
+    Owned<ABoundActivity> boundDataset = buildCachedActivity(ctx, dataset);
+    ThorActivityKind kind = TAKdiskwrite;
+    const char * activityArgName = "DiskWrite";
+    SummaryType summaryType = SummaryType::WriteFile;
+
     IHqlExpression * csvAttr = expr->queryAttribute(csvAtom);
     bool isJson = false;
     IHqlExpression * xmlAttr = expr->queryAttribute(xmlAtom);
+
     if (!xmlAttr)
     {
         xmlAttr = expr->queryAttribute(jsonAtom);
         if (xmlAttr)
-            isJson=true;
+            isJson = true;
     }
-    LinkedHqlExpr expireAttr = expr->queryAttribute(expireAtom);
-    IHqlExpression * seq = querySequence(expr);
-
-    IHqlExpression *pipe = NULL;
-    if (program)
-    {
-        if (program->getOperator()==no_pipe)
-            pipe = program->queryChild(0);
-    }
-    else if (filename->getOperator()==no_pipe)
-        pipe = filename->queryChild(0);
 
     if (pipe && expr->hasAttribute(_disallowed_Atom))
         throwError(HQLERR_PipeNotAllowed);
 
-    Owned<ABoundActivity> boundDataset = buildCachedActivity(ctx, dataset);
-    ThorActivityKind kind = TAKdiskwrite;
-    const char * activityArgName = "DiskWrite";
-    SummaryType summaryType = SummaryType::WriteFile;
     if (expr->getOperator() == no_spill)
```

Comment:
This whole if chain should probably be inside an if (useGenericReadWrites) test.  However, I would be inclined to leave that to a separate PR - this one only dealing with explicit generic formats.